File: 116_PopluatingNextRightPointerInEachNode/Solution.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    def auxConnect(self, node):
        if node.left: self.auxConnectLeft(node.left, node)
        if node.right: self.auxConnectRight(node.right, node)
        
    def auxConnectLeft(self, node, parent):
        node.next = parent.right
        self.auxConnect(node)
    
    def auxConnectRight(self, node, parent):
        if parent.next:
            node.next = parent.next.left
        else:
            node.next = None
        self.auxConnect(node)
        
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if root: 
            root.next = None
            if root.left: self.auxConnectLeft(root.left, root)
            if root.right: self.auxConnectRight(root.right, root)
        

File: 116_PopluatingNextRightPointerInEachNode/Solution.c
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  struct TreeLinkNode *left, *right, *next;
 * };
 *
 */

typedef struct TreeLinkNode TreeLinkNode;

void auxConnect(TreeLinkNode* node){
    if (node->left) auxConnectLeft(node->left, node);
    if (node->right) auxConnectRight(node->right, node);
}

void auxConnectLeft(TreeLinkNode* node, TreeLinkNode* parent) {
    node->next = parent->right;
    auxConnect(node);
}

void auxConnectRight(TreeLinkNode* node, TreeLinkNode* parent) {
    if (parent->next){
        node->next = parent->next->left;
    }
    else {
        node->next = NULL;
    }
    auxConnect(node);
}
 
void connect(struct TreeLinkNode *root) {
    if (root) {
        root->next = NULL;
        if (root->left) auxConnectLeft(root->left, root);
        if (root->right) auxConnectRight(root->right, root);
    }
}
